### ISSUE TITLE
Track feed events

### ISSUE THREAD

**Issue by benhalpern** [forem/forem#19773]
## Is this feature related to a problem? Please describe.

As a user, I want my feed to represent content I'm most likely to enjoy and engage with.

As a platform, we track net popularity, but don't track feed success rate, leading to having a harder time matching people with relevant content to them.

## Describe the solution you’d like

Begin tracking feed events for articles, similarly to how we track billboard events, ultimately allowing us to tabulate a success score.

However, we'll track success more deeply than just click-through-rate. We'll also map positive reactions and comments to feed events, to track net engagement per impression for an article's feed appearance.

We will begin by simply tabulating a success score which can be included as a factor in feed relevancy, but in the future could be very useful data in a potential evolution of the algorithm.

### Specifications

A **feed event** of category "impression" should be logged for every article card that scrolls into view in the home feed, similar to how we track billboard impressions. One impression event should be logged for every article that comes into the feed, but if possible these should be sent in batches to avoid excess pings.

A **click** should be tracked when a user clicks in from the feed.

A **reaction** should be tracked when a user makes a public positive reaction (i.e. like/unicorn, etc. and reading list) — _if_ the most recent **click** event maps to the same article.

A **comment** should also be tracked when a user comments — _if_ the most recent **click** event maps to the article.

There are edge scenarios where attribution might be dropped — for example, if I open multiple articles in multiple tabs — however this is okay. On the whole we will appropriately map the events.

`feed_success_score` should be added to the `articles` as an indexed attribute we can query and rank on. This is the tabulation of clicks + reactions + comments divided by impressions for that article. We will weight these events all evenly.

We should map each event to the current user — we do not need to track these events for anonymous sessions. This will only be fore signed in users.

We should also record which `position` in the feed each event occurs — ignoring interstitials like billboards (starting with position 1, not 0 😄)

## Definition of Done

- [ ] Create new model
- [ ] Create new summation attribute on article `feed_success_score`
- [ ] Create controller and client side tracking for impressions and clicks
- [ ] Create server-side tracking for reactions and clicks to map to feed events as per logic above

## Additional Context

This is inspired by the billboard events functionality so we should build on the patterns that work, but this is paired down in some ways (only tracking signed in, for example), and extended in others (tracking reaction and comment events)

**Comment by Ridhwana** [comment]
My initial thoughts on this issue before getting into the details is that I think this issue should be split up into smaller issues. 

> A feed event of category "impression" should be logged for every article card that scrolls into view in the home feed, similar to how we track billboard impressions. One impression event should be logged for every article that comes into the feed, but if possible these should be sent in batches to avoid excess pings.

Could be its own issue.

> A click should be tracked when a user clicks in from the feed.

Could be its own issue.
Once this one is done, we could replicate the process and implement the other two below: 

> A reaction should be tracked when a user makes a public positive reaction (i.e. like/unicorn, etc. and reading list) — if the most recent click event maps to the same article.

> A comment should also be tracked when a user comments — if the most recent click event maps to the article.

**Comment by Ridhwana** [comment]
- This is [currently how we implement the billboard impressions and clicks through JavaScript](https://github.com/forem/forem/blob/Ridhwana/tag-page-buttons/app/javascript/packs/billboardAfterRenderActions.js) and I'm assuming we may want to do something similar but we'll need to do some more extensive logic for clicks where we tracked where the click is coming from and how recent it was. 

- An additional requirement for impressions is to batch them which we could probably do in the JS before calling the endpoint.

- For impressions, we may also want to take into account that we do some checks in `initScrolling.js` to check if we're near the bottom of the page, we use a timer in this case. We may or may not want to take this into consideration but its worth at least pointing out. 

- We'll need to figure out the position of the article and exclude billboards.  We could figure this out using the DOM elements. @benhalpern Do we want to account for podcasts? 

- We'll need to add a table for the events and add endpoints for the impressions endpoint. I'm assuming that we'll need to store data like:
```
article_id, article_position, category, (maybe a context type?)
```

This table is likely to store **_a lot_** of data and so we need to think about how we plan on maintaining the data in the table. 

> We should map each event to the current user — we do not need to track these events for anonymous sessions. This will only be fore signed in users.

@benhalpern can you expand on what you mean by this please and how this affects the data that we're tracking?

**Comment by jaw6** [comment]
I agree with @Ridhwana, this should be split into separate issues, possibly with a milestone to track overall progress.

One "Naming Things" question up-front: If we want to track impressions, clicks, reactions, and comments here, I don't think we should call these `feed_events`. I'm currently leaning towards `article_events`, _however_, we have [a long-dormant initiative to rename "article" to "post"](https://github.com/forem/forem/issues/19369), so eventually we'd wind up renaming them to `post_events` ? 

It _may_ make more sense to invert naming here and name these something like `user_activities`, as it seems _likely_ that the scope here will grow above & beyond the limits of "an article", even if that's the original intent. (eg, `podcast_episodes#index`, `comments#create`, `organizations#show`, etc.)

> One impression event should be logged for every article that comes into the feed

I think this should say "comes into view" — unless we specifically want to track something like "delivered in feed but _never_ scrolled into view" ?

> if possible these should be sent in batches to avoid excess pings.

Does this refer to batching impressions of multiple articles scrolling into view together _or_ to batching impressions of the same article scrolling into view repeatedly? (I'm thinking the former, but want to confirm.)

> A click should be tracked when a user clicks in from the feed.

Is this replacing or augmenting the existing `PageView` tracking? 

"The Feed" is potentially a source of confusion here. We typically use this to refer to `stories_controller#index` (ie, and not `articles_controller#feed`, which produces an XML feed), so we might be applying this to what are, in effect, `users#show` and `organizations#show`. (Potentially, here, also `podcasts#show`, though I think that's mostly dead code now?)

It seems like we also should consider `TaggedArticles#index`, which is "feed-like", but a separate area of code, with a different load/render process (a mix of front- and back-end rendering, due to infinite scroll).

This seems like another point in favor of splitting this work into separate issues, as the work involved may not be particularly similar in practice.

> > We should map each event to the current user
> 
> Do we just want to ensure that we track the events for all logged in users or is there more to this?

My understanding here is that we should associate the event with the `current_user` if there is one, but leave it optional in case the active user is logged-out. Similar to the `belongs_to :user, optional: true` bit now in `BillboardEvent`.

**Comment by benhalpern** [comment]
> Do we want to account for podcasts?

Let's _not_ account for podcasts.

> This table is likely to store a lot of data and so we need to think about how we plan on maintaining the data in the table. Should we consider removing events after a period of time?

Good thinking — the good thing about this is that we are not attracting enough people to the home page right now, compared to general browsing, so this will still be a lot less data than, say, page views. But we will want to proactively map out a way we handle the build up.

With each row really only being a few bytes of data it shouldn't be _that_ bad to store lots of this, but we should consider our options, perhaps with a future archiving scenario of some kind.

> @benhalpern can you expand on what you mean by this please? Do we just want to ensure that we track the events for all logged in users or is there more to this?

Oh, I just mean that these should belong to the user via `user_id`. Sorry if that wasn't worded correctly.

So that would make the table:

- user_id
- article_id
- article_position
- category
- (maybe a context type?) <- And I like this train of thinking. If you can think of any other little tidbits that might be useful, I'm all for it. (Also noting that it's nice that this is such a small table, so storing lots of data isn't a huge deal).

**Comment by PhilipHow** [comment]
Split out into smaller issues.
